mitxgraders/helpers/calc.py

```diff
@@ -447,99 +324,121 @@ def parse_algebra(self):
+        expr = Forward()
 
-        # Same thing for functions
-        # Allow primes (apostrophes) at the end of function names, useful for
-        # indicating derivatives. Eg, f'(x), g''(x)
-        function = Group(inner_varname + Suppress("(") + expr + Suppress(")"))("function")
+        # Construct functions as consisting of funcname and arguments as
+        # funcname(arguments)
+        # where arguments is a comma-separated list of arguments, returned as a list
+        # Must have at least 1 argument
+        function = Group(name +
+                         Suppress("(") +
+                         Group(delimitedList(expr))("arguments") +
+                         Suppress(")")
+                         )("function")
         function.setParseAction(self.function_parse_action)
 
-        atom = number | function | varname | "(" + expr + ")"
-        atom = Group(atom)("atom")
+        # Define parentheses
+        parentheses = Group(Suppress("(") +
+                            expr +
+                            Suppress(")"))("parentheses")
+
+        # Define an atomic unit as an expression that evaluates directly to a number
+        # without the use of binary operations (assuming all children have been evaluated).
+        atom = Group(number | function | variable | parentheses)("atom")
+
+        # The following are in order of operational precedence
+        # Define exponentiation, possibly including negative powers
+        power = Group(atom + ZeroOrMore(Suppress("^") + ZeroOrMore(minus) + atom))("power")
 
-        # Do the following in the correct order to preserve order of operation.
-        pow_term = atom + ZeroOrMore("^" + atom)
-        pow_term = Group(pow_term)("power")
+        # Define negation (eg, in 5*-3 --> we need to evaluate the -3 first)
+        # Negation in powers is handled separately
+        # This has been arbitrarily assigned a higher precedence than parallel
+        negation = Group(ZeroOrMore(minus) + power)("negation")
 
-        par_term = pow_term + ZeroOrMore('||' + pow_term)  # 5k || 4k
-        par_term = Group(par_term)("parallel")
+        # Define the parallel operator 1 || 5 == 1/(1/1 + 1/5)
+        pipes = Literal('|') + Literal('|')
+        parallel = Group(negation +
+                         ZeroOrMore(Suppress(pipes) + negation))("parallel")
 
-        prod_term = par_term + ZeroOrMore((Literal('*') | Literal('/')) + par_term)  # 7 * 5 / 4
-        prod_term = Group(prod_term)("product")
+        # Define multiplication and division
+        product = Group(parallel +
+                        ZeroOrMore((Literal('*') | Literal('/')) + parallel))("product")
 
-        sum_term = Optional(plus_minus) + prod_term + ZeroOrMore(plus_minus + prod_term)  # -5 + 4 - 3
-        sum_term = Group(sum_term)("sum")
+        # Define sums and differences
+        # Note that leading - signs are treated by negation
+        sumdiff = Group(Optional(plus) + product +
```

Comment:
| Example | Expression   	| master parses?  	| multivar parses?  |
|---      |---	          |---	              |---	      |
| A        | 1+2   	      | yes   	          | yes       |
| B       | 1+-2  	      | yes  	            | yes  	    |
| C       | 1-+2  	      | yes  	            | no  	    |
| D       | 1++2  	      | yes^1             | no  	    |
| E       | 1+++2  	      | no  	            | no  	    |
| F       | ---2  	      | no  	            | yes  	    |
| G       | --+2  	      | no  	            | no  	    |

1: '++2' behaves similarly, both are parsed roughly as:
```
<sum>
  <number>1</number>
  +
  <number>
   +1 
  </number>
</sum>
```

Examples C through G are certainly a bit contrived, but I don't think they're too ridiculus.

I'm actually pretty confused about why the `ZeroOrMore` in line \#382 doesnt' permit `1+++2` (on either this branch or master).

**Suggestion:** Change `negation` to `sign` or `unary_sign`, and change `eval_negation` to `eal_sign`. (I can commit my local version of this if you want, or you can write it.)

Reply:
So, I deliberately treated + and - differently here, basing things on how google parses math input. I don't want to accept + as part of a number, like `1-+2` would require. It's ok for - to be part of a number, which is why `1-------2` works - you just keep on adding brackets around the 2 with an ever-increasing number of minus signs. 

I always thought it was strange that the old parser allowed for two pluses in a row, but not three, also. The reason is that it allowed for a single plus to be a part of a number, and then a second plus to be a part of a sum. 

I think we should discourage inputs like your D. I also don't like C, E and G for the same reason. I'm happy with B, and I'm willing to tolerate F.

Note that like 382 contains `ZeroOrMore(plus_minus + product)`, so you must have an expression after a +/- sign.

Reply:
> I don't want to accept + as part of a number
Still thinking about what you said, but to be clear my suggestion was

```python
# replace negation with this:
sign = Group(ZeroOrMore(plus_minus) + power)("sign")

# and later, replace eval_number with
def eval_sign(self, parse_result): 
        num_negations = len([char for char in parse_result[:-1] if char == '-' ])
        num = parse_result[-1]
        return num * (-1)**(num_negations)
```

Reply:
Yup, I understood the suggestion.

Do you agree with me that we shouldn't allow C/D/E/G and we need to allow B?

I'd be willing to get rid of F.

To add another case, I think `1--2` needs to be allowed, but `1---2` can probably go.

I think these can be accomplished by changing `ZeroOrMore` to `Optional` in the negation statement.

Reply:
Yes, I agree those are bad. Thinking about it more—I like rejecting them as unparseable: I think it would be easy to miscount the number of signs.

I agree Optional is a good choice.

But really, I guess it doesn't matter. I just checked some 2015 data, and in 400,000 submissions I found about 25 `+-`, 10 `-+`, and 7 `++`. (Not all of the 400,000 were FormulaResponse submissions, though. Still seems rare.)